lib/elixir/lib/calendar/date.ex

```diff
@@ -83,7 +86,7 @@ defmodule Date do
       -366
   """
 
-  @spec range(Date.t, Date.t) :: Date.Range.t
+  @spec range(Calendar.date, Calendar.date) :: Date.Range.t
```

Comment:
@josevalim I noticed a problem with this:

```elixir
# ~bad: allowing `first` be NaiveDateTime doesn't match the type of the struct which is:
# first: Date.t() (https://github.com/elixir-lang/elixir/blob/master/lib/elixir/lib/calendar/date_range.ex#L16)
iex> r = Date.range(~N[2000-01-01 09:00:00], ~D[2000-01-05])
#DateRange<~N[2000-01-01 09:00:00], ~D[2000-01-05]>

# good:
iex> Enum.to_list(r)
[~D[2000-01-01], ~D[2000-01-02], ~D[2000-01-03], ~D[2000-01-04], ~D[2000-01-05]]

# good:
iex> ~D[2000-01-01] in r
true

# bad:
iex> r.first in r
false
```

FWIW this typespec `range(Calendar.date, Calendar.date) :: Date.Range.t` matches the implementation but I should have caught the discrepancy with the `r.first in r` sooner :(

Do you think that's a problem?

I think we could:

1. keep the typespec and change implementation so it's:

   ```elixir
   iex> Date.range(~N[2000-01-01 09:00:00], ~D[2000-01-05])
   #DateRange<~N[2000-01-01], ~D[2000-01-05]>
   ```

   This is backwards incompatible

2. revert typespec to be `@spec range(Date.t, Date.t) :: Date.Range.t` and pattern match on `%Date{}` to make the function stricter. This is backwards incompatible

3. Do `2` but also handle previous code with a warning

4. Do nothing

Reply:
Let's do 2 since the range documents first and last to be dates and the behaviour is broken today.

Reply:
>  since the range documents first and last to be dates

@josevalim you mean `range/2` or `Date.Range` documents such thing? With option 1 we'll still get back a `%Date.Range{}` with first and last being `%Date{}`s, so I think we're covered there. (I edited previous comment: changed ~N to ~D in `#DateRange<~N[2000-01-01], ~D[2000-01-05]>` as it was a mistake)

I'm happy to prepare a PR for option 2! Again, I'm just not sure how we determine whether functions should be accepting `Calendar.date()` over `Date.t()`, as on recent `Date.truncate/2` discussion. If it's arbitrary I can live with that :)

Reply:
It is mostly arbitrary, yes. I would prefer to go with 2 for now, we can always relax it to be 1 later.